source/common/http/async_client_utility.h

```diff
@@ -39,5 +39,58 @@ class AsyncClientRequestTracker {
   absl::flat_hash_set<AsyncClient::Request*> active_requests_;
 };
 
+/**
+ * Sidestream watermark callback implementation for stream filter that either handles decoding only
+ * or handles both encoding and decoding.
+ */
+class StreamFilterSidestreamWatermarkCallbacks : public Http::SidestreamWatermarkCallbacks {
+public:
+  StreamFilterSidestreamWatermarkCallbacks() = default;
+
+  void onAboveWriteBufferHighWatermark() final {
```

Comment:
I'm a bit confused what this would be used for.

let's imagine a filter which sends both request and response through a side-car.  wouldn't we want to register stream callbacks for the decode path with the decode-sidecar-stream and stream callbacks for the encode path with the encode-sidecar-stream?  When do we want one object which passes "on above" to both encoder and decoder filters?

Reply:
`onAboveWriteBufferHighWatermark` only handles request path on sidestream : sidestream push back upstream or sidestream push back downstream or both.  This is due to full duplex processing .

For response path on sidestream: it will be handled in a different callback which I plan to add them in next pr (i.e. flow control part 3):) . Small prs for ease of review and scope.

More details up front: For example, when sidestream server ends back the response and downstream want to push back sidestream. This requires us to subscribe to downstream watermark events via `addDownstreamWatermarkCallbacks` This is also one of reason why I want to introduce new API. Please see point 3 in Pros here https://docs.google.com/document/d/1fwISI5wKQe6H3EOuaDLrAAgPU8Hq3KXaxJG0pm3vu-4/edit#bookmark=id.i3s841o8ul33

Reply:
ahhh, as discussed offline I didn't realize that ext_proc shared a single stream for request and response protos.   in that case this makes total sense and I think it can be adapted to handle my case as well by just having 2 StreamFilterSidestreamWatermarkCallbacks in a filter where the grpc stream wasn't shared